### PR TITLE
[Runs URI][Python] changes in deploy methods

### PR DESCRIPTION
**Pull request by sueann**
## What changes are proposed in this pull request?
 
This PR has APIs in deployment (`sagemaker`, `azureml` modules) take in the `model_uri` parameter instead of the `(model_path, run_id)` tuple. The affected APIs are:
* sagemaker.deploy
* sagemaker.run_local
* azureml.build_image
and the corresponding CLI commands.

It is very WIP. TODOs:
* Add one more test case to sagemaker tests (see inline comments for the TODO)
* Make changes to `sagemaker.run_local` and manually test
* Make changes to `azureml.__init__` file, fix tests in `tests/azureml/test_image_creation.py`
* Manually(?) test CLI changes

## How is this patch tested?

Unit tests & manual tests (fill in after they are done)
```
pytest tests/sagemaker/test_deployment.py
```

### Is this a user-facing change? 

- [ ] No. Add the `rn/none` label, then you can skip the rest of this section.
- [X] Yes. Give a description of this change to be included in the release notes for MLflow users.
 
Introduced the runs:/ URI scheme and changed APIs specifying ``(run_id, artifact_path)`` outside of MLflow Tracking to accept a single parameter ``artifact_uri`` (or ``model_uri``). The change will allow a uniform treatment of artifact locations, and enable referring to artifacts in relation to a Run in more tricky settings such as specifying artifact dependencies for a project.
 
### What component(s) does this PR affect?
 
- [ ] UI
- [X] CLI 
- [X] API 
- [ ] REST-API 
- [ ] Examples 
- [ ] Docs
- [ ] Tracking
- [ ] Projects 
- [ ] Artifacts 
- [X] Models 
- [X] Scoring 
- [X] Serving
- [ ] R
- [ ] Java
- [X] Python

### CODE REVIEW

File: mlflow/sagemaker/__init__.py [hunk 1]
```diff
@@ -318,10 +319,10 @@ def deploy(app_name, model_path, execution_role_arn=None, bucket=None, run_id=No
                     deployment_modes=",".join(DEPLOYMENT_MODES)),
                 error_code=INVALID_PARAMETER_VALUE)
 
-    s3_bucket_prefix = model_path
-    if run_id:
-        model_path = _get_model_log_dir(model_path, run_id)
-        s3_bucket_prefix = os.path.join(run_id, s3_bucket_prefix)
+    parsed = urllib.parse.urlparse(model_uri)
+    # TODO: convert to posix on windows
```
Comment (sueann) [inline mlflow/sagemaker/__init__.py]: i propose we do this later not in this PR

File: mlflow/store/runs_artifact_repo.py [hunk 2]
```diff
@@ -88,7 +88,7 @@ def list_artifacts(self, path):
 
         :return: List of artifacts as FileInfo listed directly under path.
         """
-        self.repo.list_artifacts(path)
+        return self.repo.list_artifacts(path)
```
Comment (sueann) [inline mlflow/store/runs_artifact_repo.py]: this fix is in https://github.com/mlflow/mlflow/pull/1175 so this PR doesn't have to block other PRs changing APIs with ``model_uri``

File: tests/azureml/test_image_creation.py [hunk 3]
```diff
@@ -244,6 +244,7 @@ def test_build_image_includes_default_metadata_in_azure_image_and_model_tags(skl
     with mlflow.start_run():
         mlflow.sklearn.log_model(sk_model=sklearn_model, artifact_path=artifact_path)
         run_id = mlflow.active_run().info.run_uuid
+    # TODO(sueann): how to get rid of _get_model_log_dir calls
```
Comment (sueann) [inline tests/azureml/test_image_creation.py]: should be able to just replace with `_download_artifact_from_uri`

File: tests/sagemaker/test_deployment.py [hunk 4]
```diff
@@ -2,6 +2,7 @@
 
 import os
 import pytest
+from six.moves import urllib
```
Comment (sueann) [inline tests/sagemaker/test_deployment.py]: actually don't need anymore (lint should say)

File: tests/sagemaker/test_deployment.py [hunk 5]
```diff
@@ -32,11 +34,12 @@ def pretrained_model():
     with mlflow.start_run():
         X = np.array([-2, -1, 0, 1, 2, 1]).reshape(-1, 1)
         y = np.array([0, 0, 1, 1, 1, 0])
-        lr = LogisticRegression()
+        lr = LogisticRegression(solver='lbfgs')
```
Comment (sueann) [inline tests/sagemaker/test_deployment.py]: added to get rid of the warnings when running tests

File: tests/sagemaker/test_deployment.py [hunk 6]
```diff
@@ -165,13 +165,56 @@ def test_attempting_to_deploy_in_asynchronous_mode_without_archiving_throws_exce
...
+                for config in sagemaker_client.list_endpoint_configs()["EndpointConfigs"]])
+    assert app_name in [endpoint["EndpointName"]
+                        for endpoint in sagemaker_client.list_endpoints()["Endpoints"]]
+
+
+# @mock_sagemaker_aws_services
+# def test_deploy_creates_sagemaker_and_s3_resources_with_expected_names_from_s3(
+#         pretrained_model, sagemaker_client):
+#
+#     # TODO(sueann): this case is meant to test using a remote filestore URI as the model_uri
```
Comment (sueann) [inline tests/sagemaker/test_deployment.py]: if this is tricky we can discuss whether we really need this test - there are already tests for `RunsArtifactRepository` that uses `S3ArtifactRepository` internally, but it'd be good to make sure we don't do anything weird to the URI in the deployment module.

**Review by sueann — COMMENTED** [review]
left just a few comments. thanks for adding all the tests! the test file are getting really long - maybe we can refactor them a bit later but i wouldn't worry about it now.

**Comment by sueann** [comment]
onoz conflicts!

File: mlflow/azureml/__init__.py [hunk 7]
```diff
@@ -17,7 +17,8 @@
 from mlflow.exceptions import MlflowException
 from mlflow.models import Model
 from mlflow.protos.databricks_pb2 import INVALID_PARAMETER_VALUE
-from mlflow.tracking.artifact_utils import _get_model_log_dir
+from mlflow.store.runs_artifact_repo import RunsArtifactRepository
```
[the remark this answers is not in the document]
Comment (dbczumar) [reply mlflow/azureml/__init__.py]: Not sure why lint doesn't complain. That's probably worth an investigation. Good catch!

File: mlflow/azureml/__init__.py [hunk 8]
```diff
@@ -37,10 +38,19 @@ def build_image(model_path, workspace, run_id=None, image_name=None, model_name=
...
-    :param model_path: The path to MLflow model for which the image will be built. If a run id
-                       is specified, this is should be a run-relative path. Otherwise, it
-                       should be a local path.
-    :param run_id: MLflow run ID.
+    :param model_uri: The location, in URI format, of the MLflow model for which to build an Azure
+                      ML deployment image, for example:
+
+                      - ``/Users/me/path/to/local/model``
+                      - ``relative/path/to/local/model``
+                      - ``file:///Users/me/path/to/local/model``
```
[the remark this answers is not in the document]
Comment (dbczumar) [reply mlflow/azureml/__init__.py]: Good catch! We don't. Not sure what it takes to support them beyond
```
_artifact_repository_registry.register('file', LocalArtifactRepository) 
```

Removed the example for the time being

File: mlflow/azureml/__init__.py [hunk 9]
```diff
@@ -37,10 +38,19 @@ def build_image(model_path, workspace, run_id=None, image_name=None, model_name=
...
+                      - ``s3://my_bucket/path/to/model``
+                      - ``runs://<mlflow_run_id>/path/to/model``
```
[the remark this answers is not in the document]
Comment (dbczumar) [reply mlflow/azureml/__init__.py]: Done! Thanks for catching this!

File: mlflow/sagemaker/__init__.py [hunk 10]
```diff
@@ -22,7 +22,8 @@
 from mlflow.exceptions import MlflowException
 from mlflow.models import Model
 from mlflow.protos.databricks_pb2 import RESOURCE_DOES_NOT_EXIST, INVALID_PARAMETER_VALUE
-from mlflow.tracking.artifact_utils import _get_model_log_dir
+from mlflow.store.runs_artifact_repo import RunsArtifactRepository
```
[the remark this answers is not in the document]
Comment (dbczumar) [reply mlflow/sagemaker/__init__.py]: Removed!

File: mlflow/sagemaker/__init__.py [hunk 11]
```diff
@@ -216,8 +217,19 @@ def deploy(app_name, model_path, execution_role_arn=None, bucket=None, run_id=No
...
 
     :param app_name: Name of the deployed application.
-    :param path: Path to the model. Either local if no ``run_id`` or MLflow-relative if ``run_id``
-                 is specified.
+    :param model_uri: The location, in URI format, of the MLflow model to deploy to SageMaker,
+                      for example:
+
+                      - ``/Users/me/path/to/local/model``
+                      - ``relative/path/to/local/model``
+                      - ``file:///Users/me/path/to/local/model``
```
[the remark this answers is not in the document]
Comment (dbczumar) [reply mlflow/sagemaker/__init__.py]: Removed!

File: mlflow/sagemaker/__init__.py [hunk 12]
```diff
@@ -216,8 +217,19 @@ def deploy(app_name, model_path, execution_role_arn=None, bucket=None, run_id=No
...
+                      - ``s3://my_bucket/path/to/model``
+                      - ``runs://<mlflow_run_id>/path/to/model``
```
[the remark this answers is not in the document]
Comment (dbczumar) [reply mlflow/sagemaker/__init__.py]: Done!

File: mlflow/sagemaker/__init__.py [hunk 13]
```diff
@@ -524,23 +534,31 @@ def cleanup_fn():
...
 
-    :param model_path: path to the model. Either local if no ``run_id`` or MLflow-relative if
-                                          ``run_id`` is specified.
-    :param run_id: MLflow run ID.
+    :param model_uri: The location, in URI format, of the MLflow model to serve locally,
+                      for example:
+
+                      - ``/Users/me/path/to/local/model``
+                      - ``relative/path/to/local/model``
+                      - ``file:///Users/me/path/to/local/model``
```
[the remark this answers is not in the document]
Comment (dbczumar) [reply mlflow/sagemaker/__init__.py]: Removed!

File: mlflow/sagemaker/__init__.py [hunk 14]
```diff
@@ -524,23 +534,31 @@ def cleanup_fn():
...
+                      - ``s3://my_bucket/path/to/model``
+                      - ``runs://<mlflow_run_id>/path/to/model``
```
[the remark this answers is not in the document]
Comment (dbczumar) [reply mlflow/sagemaker/__init__.py]: Fixed!

File: mlflow/utils/cli_args.py [hunk 15]
```diff
@@ -7,6 +7,13 @@
                           help="Path to the model. The path is relative to the run with the given "
                                "run-id or local filesystem path without run-id.")
 
+MODEL_URI = click.option("--model-uri", "-m", default=None, metavar="URI", required=True,
+                         help="URI to the model. This may be a local path, a 'runs://' URI, or a"
```
[the remark this answers is not in the document]
Comment (dbczumar) [reply mlflow/utils/cli_args.py]: Fixed - thanks!

File: tests/sagemaker/test_deployment.py [hunk 16]
```diff
@@ -201,20 +201,134 @@ def test_deploy_creates_sagemaker_and_s3_resources_with_expected_names(
...
+            ])
+    assert result.exit_code == 0
+
+    region_name = sagemaker_client.meta.region_name
+    s3_client = boto3.client("s3", region_name=region_name)
+    default_bucket = mfs._get_default_s3_bucket(region_name)
+    endpoint_description = sagemaker_client.describe_endpoint(EndpointName=app_name)
+    endpoint_production_variants = endpoint_description["ProductionVariants"]
+    assert len(endpoint_production_variants) == 1
+    model_name = endpoint_production_variants[0]["VariantName"]
```
[the remark this answers is not in the document]
Comment (dbczumar) [reply tests/sagemaker/test_deployment.py]: Good point! I'll create a task to follow up with a helper method later

**Review by dbczumar — COMMENTED** [review]
@sueann Addressed your comments! Thanks for the review!

File: tests/sagemaker/test_deployment.py [hunk 17]
```diff
@@ -622,8 +622,10 @@ def test_deploy_in_replace_mode_with_archiving_does_not_delete_resources(
     new_artifact_path = "model"
     with mlflow.start_run():
         mlflow.sklearn.log_model(sk_model=sk_model, artifact_path=new_artifact_path)
+        new_model_uri = "runs:/{run_id}/{artifact_path}".format(
```
Comment (sueann) [inline tests/sagemaker/test_deployment.py]: good catch!

**Review by sueann — COMMENTED** [review]
lgtm - thanks!
